talk/expert/perfectforwarding.tex

```diff
@@ -106,133 +106,179 @@
   	\begin{itemize}
-  	\item They are forbidden, but some compilers allow them
+  	\item Are formally forbidden, but can occur sometimes
     \end{itemize}
   \end{block}
   \begin{block}{}
     \begin{cppcode*}{}
       template <typename T>
-      void foo(T t) { T& k = t; } // int& &
+      void foo(T t) { T& k = t; } //int& &, error in C++03
       int ii = 4;
-      foo<int&>(ii);
+      foo<int&>(ii); // want to pass by reference
     \end{cppcode*}
   \end{block}
-  \begin{block}{\cpp11 added rvalues}
+  \begin{block}{\cpp11 added rvalue-references}
     \begin{itemize}
-    \item what about \mintinline{cpp}{int&& &} ?
-    \item and int \mintinline{cpp}{&& &&} ?
+      \item More combinations possible, like: \mintinline{cpp}{int&& &}, or \mintinline{cpp}{int&& &&}
     \end{itemize}
   \end{block}
-  \begin{exampleblock}{Rule}
-    \mintinline{cpp}{&} always wins\\
-    \mintinline{cpp}{&& &, & &&, & &} $\rightarrow$ \mintinline{cpp}{&}\\
-    \mintinline{cpp}{&& &&} $\rightarrow$ \mintinline{cpp}{&&}
+  \begin{exampleblock}{Reference collapsing}
+    \begin{itemize}
+      \item Rule: When multiple references are involved, \mintinline{cpp}{&} always wins
+      \item \mintinline{cpp}{T&& &, T& &&, T& &} $\rightarrow$ \mintinline{cpp}{T&}
+      \item \mintinline{cpp}{T&& &&} $\rightarrow$ \mintinline{cpp}{T&&}
+    \end{itemize}
   \end{exampleblock}
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[11]{rvalue in type-deducing context}
-  \begin{block}{}
+  \frametitlecpp[11]{Forwarding references}
+  \begin{exampleblock}{Rvalue reference in type-deducing context}
     \begin{cppcode*}{}
       template <typename T>
-      void func(T&& t) {}
+      void f(T&& t) { ... }
     \end{cppcode*}
-  \end{block}
-  \begin{itemize}
-    \item Next to a template parameter, \mintinline{cpp}{&&} is not an rvalue, but a ``forwarding reference'' (aka.\ ``universal reference'')
-    \item \mintinline{cpp}{T&&} actual type depends on the arguments passed to func
+  \end{exampleblock}
+  \begin{block}{Forwarding reference}
     \begin{itemize}
-      \item if an lvalue of type U is given, T is deduced to \mintinline{cpp}{U&}
-      \item otherwise, collapse references normally
+      \item Next to a template parameter, that can be deduced, \mintinline{cpp}{&&} is not an rvalue reference, but a ``forwarding reference'' (aka.\ ``universal reference'')
+      \begin{itemize}
+      	\item So, this applies only to functions
+      \end{itemize}
+      \item The template parameter is additionally allowed to be deduced as an lvalue reference, and reference collapsing proceeds:
+      \begin{itemize}
+        \item if an lvalue of type \mintinline{cpp}{U} is given, \mintinline{cpp}{T} is deduced as \mintinline{cpp}{U&} and the parameter type \mintinline{cpp}{U& &&} collapses to \mintinline{cpp}{U&}
+        \item otherwise (rvalue), T is deduced as \mintinline{cpp}{U} and forms the parameter type \mintinline{cpp}{U&&}
```

Comment:
```suggestion
        \item otherwise (rvalue), T is deduced as \mintinline{cpp}{U}
        \begin{itemize}
          \item and forms the parameter type \mintinline{cpp}{U&&}
        \end{itemize}
```